### PR TITLE
[API] As a user, I can query all the stored data

### LINKED ISSUE #22: [API] As a user, I can query all the stored data
### Why?
User can make an API request for all stored data (10 records/page) 

### Acceptance Criteria
- [x] Provide an endpoint to query all stored data
- [x] Return all stored data
- [x] Paginating those data 10 records/page

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/22

## What happened 👀

The user can make a request to the endpoint `[GET] /api/v1/keywords` to query all the stored data belonging to the user.
And also filter with `keyword` and `page` ✌️

## Insight 📝

- [x] Provide an endpoint to query all stored data
- [x] Return all stored data
- [x] Paginating those data 10 records/page

Sample request:

```curl
curl --location --request GET 'http://localhost:8080/api/v1/keywords?keyword=nd&p=1' \
--header 'Authorization: Bearer NWRHMTEXZTMTZMY0OS0ZZDKYLTHLMZMTMZIWNMNHNDDLYTC4'
```

Extra:

Providing valid JSON-LD format based on this schema here https://jsonapi.org/examples/#pagination

## Proof Of Work 📹

[GET] /api/v1/keywords?[keyword=XXXX]&[p=XXXX]
Query all stored data
| Valid Request | Invalid Request |
|---------------|----------------|
| Screen Shot 2564-04-02 at 23 35 52 | Screen Shot 2564-04-02 at 23 36 22 |

### CODE REVIEW

File: controllers/api/v1/keyword_test.go [hunk 1]
```diff
@@ -24,6 +27,119 @@ var _ = Describe("KeywordController", func() {
...
+				accessToken := FabricatorAccessToken(oauthClient.ID, user.Id)
+				headers := http.Header{"Authorization": {fmt.Sprintf("Bearer %v", accessToken)}}
+
+				response := MakeAuthenticatedRequest("GET", "/api/v1/keywords", headers, nil, nil)
+
+				Expect(response).To(MatchJSONSchema("keyword/list/valid"))
+			})
+
+			Context("given `p`(page) param is set", func() {
+				It("shows only records in page number 2", func() {
```
Comment (junan) [inline controllers/api/v1/keyword_test.go]:
```suggestion
				It("shows only records of page number 2", func() {
```

File: helpers/utility_map_helper_test.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	"go-crawler-challenge/helpers"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("UtilityMapHelper", func() {
+	Describe("#GetMapKeyByNumber", func() {
+		Context("given a valid map list with integer", func() {
+			It("returns a firstly key", func() {
```
Comment (junan) [inline helpers/utility_map_helper_test.go]:
```suggestion
			It("returns first key", func() {
```

File: controllers/api/base.go [hunk 3]
```diff
@@ -40,23 +42,58 @@ func (c *BaseController) Prepare() {
...
+}
+
 func (c *BaseController) MappingPolicy(method string, policy Policy) {
 	c.actionPolicy[method] = policy
 }
 
-func (c *BaseController) RenderJSON(data interface{}, status int) {
-	c.Ctx.Output.Header("Content-Type", ContentType)
-	c.Ctx.ResponseWriter.WriteHeader(status)
+func (c *BaseController) RenderJSONMany(data interface{}, meta *jsonapi.Meta, links *jsonapi.Links, status int) {
```
Comment (junan) [inline controllers/api/base.go]: `RenderJSONMany` name is odd to me 🤔

Comment (Lahphim) [reply controllers/api/base.go]: @junan I make it consistences with this https://github.com/google/jsonapi/blob/master/node.go#L25
But I'll change to `RenderJSONList`

**Review by junan — APPROVED** [review]

File: helpers/utility_map_helper.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package helpers
+
+// GetMapKeyByNumber returns a firstly key from searching by value
```
Comment (olivierobert) [inline helpers/utility_map_helper.go]: English:

```suggestion
// GetMapKeyByNumber returns the key of the first entry matched by its value
```

File: helpers/utility_map_helper.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+func GetMapKeyByNumber(mapList map[string]int, value int) string {
```
Comment (olivierobert) [inline helpers/utility_map_helper.go]: How about making it more explicit:

```suggestion
func GetMapKeyByNumberValue(mapList map[string]int, value int) string {
```

File: controllers/api/constants.go [hunk 6]
```diff
@@ -4,10 +4,17 @@ import (
...
+)
 
 var (
-	ErrorInvalidUser        = errors.New("invalid user")
-	ErrorMissingAccessToken = errors.New("missing access token")
-	ErrorNotFoundUser       = errors.New("user not found")
+	DefaultOrderBy = []string{"created_at desc"}
+
+	ErrorInvalidUser           = errors.New("invalid user")
+	ErrorInvalidPayloaderType  = errors.New("invalid payloader type")
```
Comment (olivierobert) [inline controllers/api/constants.go]: Did you mean:

```suggestion
	ErrorInvalidPayloadType  = errors.New("invalid payload type")
```

Note sure what would be `payloader` 😅

File: controllers/api/v1/keyword.go [hunk 7]
```diff
@@ -20,14 +24,56 @@ func (c *KeywordController) NestPrepare() {
...
+	queryList := map[string]interface{}{
+		"user_id":            c.CurrentUser.Id,
+		"keyword__icontains": keyword,
+	}
+
+	totalRows, err := models.CountAllKeyword(queryList)
+	if err != nil {
+		c.RenderGenericError(ErrorRetrieveKeywordFailed)
+
+		return
```
Comment (hoangmirs) [inline controllers/api/v1/keyword.go]: I think we can remove the `return` here as we're using `StopRun()` in `renderError()`, right?

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**